src/main/java/org/ibp/api/java/impl/middleware/common/validator/AttributeValidator.java

```diff
@@ -45,88 +51,99 @@ public void validateAttributeIds(final BindingResult errors, final List<String>
 		final List<Integer> ids =
 			attributeIds.stream().map(attributeId -> Integer.valueOf(attributeId)).distinct().collect(Collectors.toList());
 
-		final List<Attribute> attributes = this.germplasmDataManager.getAttributeByIds(ids);
-		if (attributes.size() != ids.size()) {
+		final VariableFilter variableFilter = new VariableFilter();
+		ids.forEach(variableFilter::addVariableId);
+
+		final List<Variable> variables = this.ontologyVariableDataManager.getWithFilter(variableFilter);
+		final boolean anyVariableWithNoAllowedType =
+			variables.stream().anyMatch(v -> !v.getVariableTypes().contains(VariableType.GERMPLASM_ATTRIBUTE) &&
+				!v.getVariableTypes().contains(VariableType.GERMPLASM_PASSPORT));
+		if (variables.size() != ids.size() || anyVariableWithNoAllowedType) {
 			errors.reject("attribute.invalid", "");
 		}
 	}
 
-	public void validateAttributeType(final BindingResult errors, final String attributeType) {
-		if (!StringUtils.isEmpty(attributeType) && !AttributeValidator.ALLOWED_ATTRIBUTE_TYPES.contains(attributeType.toUpperCase())) {
-			errors.reject("attribute.type.invalid", "");
+	public void validateAttributeVariable(final BindingResult errors, final Variable variable) {
+		if (variable == null) {
+			errors.reject("attribute.variable.invalid", "");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+		if (!CollectionUtils.containsAny(variable.getVariableTypes(), ALLOWED_ATTRIBUTE_TYPES)) {
+			errors.reject("attribute.variable.type.invalid", "");
 			throw new ApiRequestValidationException(errors.getAllErrors());
 		}
 	}
 
-	void validateAttributeCode(final BindingResult errors, final GermplasmAttributeRequestDto dto) {
-		final List<AttributeDTO> attributeDTOS = this.germplasmAttributeService
-			.filterGermplasmAttributes(Collections.singleton(dto.getAttributeCode()), dto.getAttributeType());
-		if(attributeDTOS.isEmpty()) {
-			errors.reject("attribute.code.invalid", "");
+	public void validateVariableDataTypeValue(final BindingResult errors, final Variable variable, final String value) {
+		if (!VariableValueUtil.isValidAttributeValue(variable, value)) {
+			errors.reject("invalid.variable.value", "");
 			throw new ApiRequestValidationException(errors.getAllErrors());
 		}
 	}
 
 	void validateGermplasmAttributeShouldNotExist(final BindingResult errors, final Integer gid, final GermplasmAttributeRequestDto dto) {
-		final List<GermplasmAttributeDto> germplasmAttributeDtos = this.germplasmAttributeService.getGermplasmAttributeDtos(gid,
-			dto.getAttributeType());
-
+		//FIXME Search by gid and variableId, to be addressed in IBP-4659
```

Comment:
Same as in Middleware comment: IBP-4659 is marked resolved, is there another tech backlog ticket?